app/Http/Controllers/Auth/OAuth2Controller.php

```diff
@@ -0,0 +1,203 @@
+     */
+    private $auth;
+
+    /**
+     * @var \Pterodactyl\Contracts\Repository\UserRepositoryInterface
+     */
+    protected $repository;
+
+    /**
+     * @var \Pterodactyl\Services\Users\UserUpdateService
+     */
+    protected $updateService;
+
+    /**
+     * @var \Prologue\Alerts\AlertsMessageBag
+     */
+    protected $alert;
+
+    /**
+     * OAuth2Controller constructor.
+     *
+     * @param \Pterodactyl\Services\Users\UserCreationService $creationService
+     * @param \Illuminate\Contracts\Hashing\Hasher $hasher
+     * @param \Illuminate\Auth\AuthManager $auth
+     * @param \Pterodactyl\Contracts\Repository\UserRepositoryInterface $repository
+     * @param \Pterodactyl\Services\Users\UserUpdateService $updateService
+     * @param AlertsMessageBag $alert
+     */
+    public function __construct(UserCreationService $creationService, Hasher $hasher, AuthManager $auth, UserRepositoryInterface $repository, UserUpdateService $updateService, AlertsMessageBag $alert)
+    {
+        $this->creationService = $creationService;
+        $this->hasher = $hasher;
+        $this->auth = $auth;
+        $this->repository = $repository;
+        $this->updateService = $updateService;
+        $this->alert = $alert;
+    }
+
+    public function login()
+    {
+        // If this feature is not enabled or the user is already signed in and isn't linking a new account redirect or abort
+        if ($this->auth->guard()->check() and ! session()->has('link_oauth2_driver')) {
+            return redirect()->route('index');
+        }
+        if (! config('oauth2.enabled')) {
+            abort(404);
+        }
+
+        // If linking a new provider form the account page
+        if (session()->has('link_oauth2_driver')) {
+            $session_driver = session()->get('link_oauth2_driver');
+            session()->forget('link_oauth2_driver');
+
+            // Check if authenticated
+            if (! $this->auth->guard()->check()) {
+                $errors = new MessageBag(['user' => [__('auth.failed')]]);
+
+                return redirect()->route('auth.login')
+                    ->withErrors($errors);
+            }
+
+            $user = $this->auth->guard()->user();
+
+            // Get the user info from OAuth2
+            try {
+                $socialite_user = Socialite::driver($session_driver)->user();
+            } catch (\Exception $exception) {
+                $this->alert->danger(trans('base.account.oauth2_link_failed'))->flash();
+
+                return redirect()->route('account');
+            }
+
+            $oauth2_socialite_id = $socialite_user->getId();
+
+            $new_ids = [];
+            $done = false;
+            // Replace if already exists
+            foreach (preg_split('~,~', $user->getAttributes()['oauth2_id']) as $id) {
+                if (Str::startsWith($id, $session_driver)) {
+                    $id = $session_driver . ':' . $oauth2_socialite_id;
+                    $done = true;
```

Comment:
Should can break out of the loop at this point most likely, or is there a reason to continue through the remaining?